modules/storages/lib/api/v3/file_links/storage_url_helper.rb

```diff
@@ -33,4 +33,65 @@ def storage_url_open(file_link, open_location: false)
 
     "#{file_link.storage.host}/f/#{file_link.origin_id}?openfile=#{location_flag}"
   end
+
+  # rubocop:disable Metrics/AbcSize
+  def make_download_url(file_link:, user:)
+    storage = file_link.storage
+    oauth_client = storage.oauth_client
+
+    client_token = ::OAuthClients::ConnectionManager
+                     .new(user:, oauth_client:)
+                     .get_access_token
+
+    return ServiceResult.failure(result: I18n.t('http.request.missing_authorization')) if client_token.failure?
+
+    response = make_direct_download oauth_client:,
+                                    access_token: client_token.result.access_token,
+                                    file_id: file_link.origin_id
+
+    return response if response.failure?
+
+    token = parse_direct_download_token body: response.result
+    if token.blank?
+      Rails.logger.error "Received unexpected json response: #{response.result}"
+      return ServiceResult.failure(result: I18n.t('http.response.unexpected'))
+    end
+
+    url = "#{storage.host}/apps/integration_openproject/direct/#{token}/#{file_link.origin_name}"
+    ServiceResult.success(result: url)
+  end
+
+  # rubocop:enable Metrics/AbcSize
+
+  private
+
+  def make_direct_download(oauth_client:, access_token:, file_id:)
+    response = API::V3::Storages::StorageRequestFactory
+                 .new(oauth_client:)
+                 .download_command
+                 .call(access_token:, file_id:)
+
+    return response if response.failure?
+
+    # The nextcloud API returns a successful response with empty body if the authorization is missing or expired
+    return ServiceResult.failure(result: I18n.t('http.request.failed_authorization')) if response.result.body.blank?
```

Comment:
Same with this method, this could be part of the above mentioned service

Reply:
you mean, this method of 12 lines is too complex? Or do you refer to another issue with `same with this method`?

Reply:
"Same" meaning that no it's not too complex, but should also be part of the service

Reply:
So, a service only for making download URLs? As this method is a private one, it would of course move along the caller.

I didn't really find in code, how the S3 urls are made for attachments.

Reply:
S3 urls are coming from the fog-aws library, so we don't build it ourselves

Reply:
resolved, as no service extraction necessary for now.